piet-coregraphics/src/text.rs

```diff
@@ -0,0 +1,433 @@
+//! Text related stuff for the coregraphics backend
+
+use std::marker::PhantomData;
+
+use core_foundation_sys::base::CFRange;
+use core_graphics::base::CGFloat;
+use core_graphics::context::CGContextRef;
+use core_graphics::geometry::{CGPoint, CGRect, CGSize};
+use core_graphics::path::CGPath;
+use core_text::font::{self, CTFont};
+
+use piet::kurbo::{Point, Size};
+use piet::{
+    Error, Font, FontBuilder, HitTestMetrics, HitTestPoint, HitTestTextPosition, LineMetric, Text,
+    TextLayout, TextLayoutBuilder,
+};
+
+use crate::ct_helpers::{AttributedString, Frame, Framesetter, Line};
+
+// inner is an nsfont.
+#[derive(Debug, Clone)]
+pub struct CoreGraphicsFont(CTFont);
+
+pub struct CoreGraphicsFontBuilder(Option<CTFont>);
+
+#[derive(Clone)]
+pub struct CoreGraphicsTextLayout {
+    string: String,
+    attr_string: AttributedString,
+    framesetter: Framesetter,
+    pub(crate) frame: Option<Frame>,
+    // distance from the top of the frame to the baseline of each line
+    pub(crate) line_y_positions: Vec<f64>,
+    /// offsets in utf8 of lines
+    line_offsets: Vec<usize>,
+    pub(crate) frame_size: Size,
```

Comment:
Note to self while reviewing: should this be a `Rect` rather than a `Size`, or can we always assume the origin is at zero?

Reply:
this is a size. It's hard to even say what origin means here, but in some sense it is at zero. (the drawing origin is the baseline of the first line, but we aren't given it explicitly)